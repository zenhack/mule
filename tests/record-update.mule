#XFAIL: type formatting
{ x = {} } where { y = { z = {} } }
