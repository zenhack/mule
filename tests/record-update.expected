
Desugared:

	(((.= y) (letrec (types ()) (values ((x {}))) (((.= x) {}) x)))
	 (letrec (types ()) (values ((z {}))) (((.= z) {}) z)))

inferred type:

	(all a
	 (all b
	  (all c
	   (all d
	    (record (types (...d))
	     (values
	      ((x (record (types (...c)) (values ())))
	       (y
	        (record (types (...b))
	         (values ((z (record (types (...a)) (values ()))))))))))))))

Runtime term:

	(App
	 ((App
	   ((Lam
	     (0 () (Lam (1 () (Update (old (Var 1)) (label y) (field (Var 0)))))))
	    (LetRec
	     (((0 (Record ())))
	      (App
	       ((App
	         ((Lam
	           (0 ()
	            (Lam (1 () (Update (old (Var 1)) (label x) (field (Var 0)))))))
	          (Record ())))
	        (Var 0)))))))
	  (LetRec
	   (((0 (Record ())))
	    (App
	     ((App
	       ((Lam
	         (0 () (Lam (1 () (Update (old (Var 1)) (label z) (field (Var 0)))))))
	        (Record ())))
	      (Var 0)))))))

Evaluated:

	(Record ((x (Record ())) (y (Record ((z (Record ())))))))
